master/buildbot/process/build.py

```diff
@@ -291,8 +283,59 @@ def startBuild(self, build_status, workerforbuilder):
             log.err(Failure())
             self.buildFinished(['Build.setupBuild', 'failed'], EXCEPTION)
             return
+
         # flush properties in the beginning of the build
         yield self._flushProperties(None)
+
+        try:
+            ready = yield workerforbuilder.prepare(self)
+        except Exception:
+            log.err(Failure(), 'while preparing workerforbuilder:')
+            self.buildFinished(["worker", "not", "available"], RETRY)
+            return
+
+        # If prepare returns True then it is ready and we start a build
+        # If it returns false then we don't start a new build.
+        if not ready:
+            log.msg("worker %s can't build %s after all; retrying the "
+                    "build" % (self, workerforbuilder))
+            self.stopped = True
+            self.buildFinished(["worker", "not", "available"], RETRY)
```

Comment:
Well I am a little embarrassed with such reasoning of "it is not needed with current implementation", and usually prefer to just always wait for all the asynchronous stuff that we call.
It is so hard to debug when there are race conditions because of unwaited deferred.

Its more like a coding style rule. You shalt always wait for the deferred, and don't try to be smart at it.
Another additional remark on that topic is that I am not sure if SynchronousTestCase do help catching unwaited deferred.